# Kaggle exploration notebook

This notebook aims to explore the data downloaded from kaggle directly in order to understand the data. The main idea is to merge files in order to have clean csv files to work with the API.

# Table of contents

- [Train analysis](#train-analysis)
- [Transactions analysis](#transactions-analysis)
- [Stores analysis](#stores-analysis)
- [Oil analysis](#oil-analysis)
- [Holidays analysis](#holidays-analysis)

Let's start by importing our python modules.

In [1]:
import pandas as pd

from retail_api.core.paths import KAGGLE_DIR
import numpy as np

Then, we can load the CSV files. We are obimitting `sample_submission.csv` and `test.csv` because the aim is not to participate at the Kaggle competion.

In [2]:
holidays_df = pd.read_csv(KAGGLE_DIR / "holidays_events.csv", parse_dates=["date"])
oil_df = pd.read_csv(KAGGLE_DIR / "oil.csv", parse_dates=["date"])
stores_df = pd.read_csv(KAGGLE_DIR / "stores.csv")
transactions_df = pd.read_csv(KAGGLE_DIR / "transactions.csv", parse_dates=["date"])
train_df = pd.read_csv(KAGGLE_DIR / "train.csv", parse_dates=["date"])

# Train analysis

Now, we can analyze the main information contained in the `train.csv` file which is the main one.

In [3]:
train_df.head()

,id,date,store_nbr,family,sales,onpromotion
0,0,2013-01-01,1,AUTOMOTIVE,0.0,0
1,1,2013-01-01,1,BABY CARE,0.0,0
2,2,2013-01-01,1,BEAUTY,0.0,0
3,3,2013-01-01,1,BEVERAGES,0.0,0
4,4,2013-01-01,1,BOOKS,0.0,0


In [4]:
print(f"{train_df.columns} \n {train_df.dtypes} \n {train_df.shape}")

Index(['id', 'date', 'store_nbr', 'family', 'sales', 'onpromotion'], dtype='str') 
 id                      int64
date           datetime64[us]
store_nbr               int64
family                    str
sales                 float64
onpromotion             int64
dtype: object 
 (3000888, 6)


In [5]:
min_date = train_df["date"].min()
print(f"Min date in train set: {min_date}")

max_date = train_df["date"].max()
print(f"Max date in train set: {max_date}")

Min date in train set: 2013-01-01 00:00:00
Max date in train set: 2017-08-15 00:00:00


In [6]:
train_df["date"].nunique()

1684

In [7]:
train_df["store_nbr"].nunique()

54

In [8]:
train_df["family"].nunique()

33

The dataset contains daily aggregated sales observations at the store and product family level.

More precisely, it runs over **54 stores** over the period from **2013-01-01 to 2017-08-15** on **1684 days**. There are **33 product families** available in each store. By multiplying the number of stores, the number of days, and the number of product families, we obtain the total number of rows in the dataset, which is **3000888**.

The `sales` column represents the total sales value for the given store, product family and day. **Fractional values are possible** since products can be sold in fractional units (1.5 kg of cheese, for instance, as opposed to 1 bag of chips). The `onpromotion` column indicates how many items from this product family were under promotion on that day in the given store.

As the first objective of the project is to build an API simulating a daily data flow, we will not push further the analysis of the data. The goal at this point is just understanding the data structure and not to perform an in-depth business or statistical analysis.

[Back to top](#kaggle-exploration-notebook)

# Transactions analysis

In this section, we will analyze the `transactions.csv` file.

In [9]:
transactions_df.head(5)

,date,store_nbr,transactions
0,2013-01-01,25,770
1,2013-01-02,1,2111
2,2013-01-02,2,2358
3,2013-01-02,3,3487
4,2013-01-02,4,1922


In [10]:
print(f"{transactions_df.columns} \n {transactions_df.dtypes} \n {transactions_df.shape}")

Index(['date', 'store_nbr', 'transactions'], dtype='str') 
 date            datetime64[us]
store_nbr                int64
transactions             int64
dtype: object 
 (83488, 3)


In [11]:
min_date = transactions_df["date"].min()
print(f"Min date in train set: {min_date}")

max_date = transactions_df["date"].max()
print(f"Max date in train set: {max_date}")

Min date in train set: 2013-01-01 00:00:00
Max date in train set: 2017-08-15 00:00:00


In [12]:
transactions_df["date"].nunique()

1682

In [13]:
transactions_df["store_nbr"].nunique()

54

In [14]:
transactions_df.groupby("store_nbr")["date"].count().describe()

count      54.000000
mean     1546.074074
std       331.899101
min       118.000000
25%      1659.250000
50%      1676.000000
75%      1677.000000
max      1678.000000
Name: date, dtype: float64

The `transactions.csv` file contains the number of daily transactions for each store. All the **54** stores are represented. The overall time span is similar to the one observed in the train.csv dataset.

We observe **1682** unique days in this dataset over the period from **2013-01-01 to 2017-08-15** on **1684 days**, which is slightly lower than the number of days represented in the `train.csv` file. Moreover, the store which has the lower amount of recorded days of transactions contains only **188** recorded days, and the higher **1678**. In other words, there are missing transactions values for all the stores.

In average, each store has recorded transactions over **1546** days. Regarding the quantiles, we observe that the majority of the stores contains almost all the dates, and that a few of them contains a very low amount of dates, significantly reducing the overall average.

[Back to top](#kaggle-exploration-notebook)

# Stores analysis

We are now performing the analysis of the `stores.csv` file.

In [15]:
stores_df.head(5)


,store_nbr,city,state,type,cluster
0,1,Quito,Pichincha,D,13
1,2,Quito,Pichincha,D,13
2,3,Quito,Pichincha,D,8
3,4,Quito,Pichincha,D,9
4,5,Santo Domingo,Santo Domingo de los Tsachilas,D,4


In [16]:
print(f"{stores_df.columns} \n {stores_df.dtypes} \n {stores_df.shape}")

Index(['store_nbr', 'city', 'state', 'type', 'cluster'], dtype='str') 
 store_nbr    int64
city           str
state          str
type           str
cluster      int64
dtype: object 
 (54, 5)


In [17]:
stores_df["store_nbr"].nunique()

54

In [18]:
stores_df["city"].nunique()

22

In [19]:
stores_df["state"].nunique()

16

In [20]:
stores_df["type"].nunique()

5

In [21]:
stores_df["cluster"].nunique()

17

In [22]:
np.sort(stores_df["cluster"].unique())

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17])

In [23]:
stores_df.groupby("type")["cluster"].unique()

type
A            [5, 11, 14, 17]
B                [6, 16, 10]
C             [15, 7, 3, 12]
D    [13, 8, 9, 4, 1, 10, 2]
E                       [10]
Name: cluster, dtype: object

The `stores.csv` file provides descriptive information for the **54** stores included in the dataset. For each store, it includes the **city**, **state**, **store type**, and **cluster identifier**.

The **54 stores** are located in **22 distinct cities** across **16 states**, providing a broad geographical coverage of the retail network.

The **city** and **state** information are written in full. The **type** variable is encoded as a categorical letter, while the **cluster** variable is represented by an integer identifier. According to the documentation, the **cluster** variable groups stores with similar characteristics or commercial behaviour.

We observe that **cluster identifiers** are not strictly associated with a single store type. In particular, **cluster 10** appears across several store types (B, D and E). This suggests that the cluster segmentation captures characteristics that are not directly related to the store type classification. This indicates that store type and cluster should be considered as complementary categorical variables rather than hierarchical groupings.

[Back to top](#kaggle-exploration-notebook)

# Oil analysis

In [24]:
oil_df.head(5)

,date,dcoilwtico
0,2013-01-01,NaN
1,2013-01-02,93.14
2,2013-01-03,92.97
3,2013-01-04,93.12
4,2013-01-07,93.20


In [25]:
print(f"{oil_df.columns} \n {oil_df.dtypes} \n {oil_df.shape}")

Index(['date', 'dcoilwtico'], dtype='str') 
 date          datetime64[us]
dcoilwtico           float64
dtype: object 
 (1218, 2)


In [26]:
min_date = oil_df["date"].min()
print(f"Min date in train set: {min_date}")

max_date = oil_df["date"].max()
print(f"Max date in train set: {max_date}")

Min date in train set: 2013-01-01 00:00:00
Max date in train set: 2017-08-31 00:00:00


In [27]:
min_price = oil_df["dcoilwtico"].min()
print(f"Min price in train set: {min_price}")

max_price = oil_df["dcoilwtico"].max()
print(f"Max price in train set: {max_price}")

Min price in train set: 26.19
Max price in train set: 110.62


In [28]:
oil_df["dcoilwtico"].describe()

count    1175.000000
mean       67.714366
std        25.630476
min        26.190000
25%        46.405000
50%        53.190000
75%        95.660000
max       110.620000
Name: dcoilwtico, dtype: float64

In [29]:
oil_df["dcoilwtico"].isnull().sum()

np.int64(43)

The `oil.csv` file contains daily oil price observations over the period from 2013-01-01 to 2017-08-31, with a total of 1218 entries.

The price ranges from 26.19 up to 110.62 during the concerned period, with a mean price of 67.71 and a median price of 53.19. The median being lower than the mean suggests a distribution with a few relatively high price values.

We observe that there are 43 missing values in the dataset and that observations are not available for every calendar day.

[Back to top](#kaggle-exploration-notebook)

# Holidays analysis

In [30]:
holidays_df.head(5)

,date,type,locale,locale_name,description,transferred
0,2012-03-02,Holiday,Local,Manta,Fundacion de Manta,False
1,2012-04-01,Holiday,Regional,Cotopaxi,Provincializacion de Cotopaxi,False
2,2012-04-12,Holiday,Local,Cuenca,Fundacion de Cuenca,False
3,2012-04-14,Holiday,Local,Libertad,Cantonizacion de Libertad,False
4,2012-04-21,Holiday,Local,Riobamba,Cantonizacion de Riobamba,False


In [31]:
print(f"{holidays_df.columns} \n {holidays_df.dtypes} \n {holidays_df.shape}")

Index(['date', 'type', 'locale', 'locale_name', 'description', 'transferred'], dtype='str') 
 date           datetime64[us]
type                      str
locale                    str
locale_name               str
description               str
transferred              bool
dtype: object 
 (350, 6)


In [32]:
min_date = holidays_df["date"].min()
print(f"Min date in holydays set: {min_date}")

max_date = holidays_df["date"].max()
print(f"Max date in holydays set: {max_date}")

Min date in holydays set: 2012-03-02 00:00:00
Max date in holydays set: 2017-12-26 00:00:00


In [33]:
print(f"There is {holidays_df['type'].nunique()} different holiday types which are:")
print(holidays_df["type"].unique())

There is 6 different holiday types which are:
<StringArray>
['Holiday', 'Transfer', 'Additional', 'Bridge', 'Work Day', 'Event']
Length: 6, dtype: str


In [34]:
holidays_df["locale"].unique()

<StringArray>
['Local', 'Regional', 'National']
Length: 3, dtype: str

In [35]:
holidays_df["locale_name"].unique()

<StringArray>
[                         'Manta',                       'Cotopaxi',
                         'Cuenca',                       'Libertad',
                       'Riobamba',                           'Puyo',
                       'Guaranda',                       'Imbabura',
                      'Latacunga',                        'Machala',
                  'Santo Domingo',                      'El Carmen',
                        'Cayambe',                     'Esmeraldas',
                        'Ecuador',                         'Ambato',
                         'Ibarra',                        'Quevedo',
 'Santo Domingo de los Tsachilas',                    'Santa Elena',
                          'Quito',                           'Loja',
                        'Salinas',                      'Guayaquil']
Length: 24, dtype: str

In [36]:
holidays_df["description"].nunique()

103

In [37]:
holidays_df["transferred"].unique()

array([False,  True])

In [38]:
holidays_df[["transferred"]][holidays_df["transferred"] == True].count()

transferred    12
dtype: int64

The `holidays_events.csv` file contains contains calendar information about public holidays and special events on the period from 2012-03-02 to 2017-12-26. There is a total of 350 records.

There is 6 different holidays types which are 'Holiday', 'Transfer', 'Additional', 'Bridge', 'Work Day', 'Event'.
- 'Holiday' refers to an official public holiday
- 'Transfer' indicates that a official public holiday has been officially moved to this date.
- 'Additional' indicates an extra holiday
- 'Bridge' indicates a bridging day between a holiday and a weekend
- 'Work day' indicates a day that would normally be a holiday but has been designated as a working day.
- 'Event' indicates a day that would normally be a holiday but has been designated as a working day.

The variable **locale** indicates if the holiday is rather local to the city, the region, or is national. Therefore, the **locale_name** variable tells in which locality where the holiday applies.

The variable **description** is a textual description of the observed holiday.

Finally, the **transferred** boolean variable indicates whether a holiday has been moved to another date. When set to True, the original date is treated as a working day, while a corresponding entry with type 'Transfer' indicates the new observed holiday date.

[Back to top](#kaggle-exploration-notebook)